src/main/java/org/variantsync/diffdetective/variationtree/VariationNode.java

```diff
@@ -0,0 +1,583 @@
+ * prominently, there are no methods to change the {@link getNodeType type} or the
+ * {@link getLabelLines label} of this node.
+ *
+ * <p>There are many methods which are not abstract. These are convenience methods or algorithms
+ * acting on variation nodes where a separate class may be undesirable (for example because they are
+ * quite common or because the calling syntax {@code node.algorithm()} makes more sense than the
+ * alternative {@code Algorithm.run(node)}).
+ *
+ * <p>Identity of {@code VariationNode}s shouldn't be tested using {@code ==} because this might be
+ * a view which is instantiated multiple times for the same node (e.g., it might be that
+ * {@code getParent() != getParent()}). Instead the method {@link isSameAs} should be used to test
+ * for identity.
+ *
+ * @param <T> the backing type storing the actual information presented by this class. This is
+ * currently either a {@link VariationTreeNode concrete variation tree} (in this case
+ * {@link getBackingNode} acts as identity) or a {@link DiffNode.Projection projection} of a
+ * {@link DiffNode}.
+ *
+ * @see assertConsistency
+ * @author Benjamin Moosherr
+ */
+public abstract class VariationNode<T> implements HasNodeType {
+    /**
+     * Returns the class instance storing the actual attributes of this node. This currently is
+     * either a {@link VariationTreeNode concrete variation tree} or the projection of a
+     * {@link DiffNode} at a specific time.
+     */
+    public abstract T getBackingNode();
+
+    /**
+     * Returns the node type of this node which determines the type of the represented element in
+     * the variation tree (e.g., mapping or artifact).
+     *
+     * @see HasNodeType
+     */
+    public abstract NodeType getNodeType();
+
+    /**
+     * Returns the label of this node as an unmodifiable list of lines.
+     *
+     * <p>If {@link isArtifact} is {@code true} this may represent the source code of this artifact.
+     * Otherwise it may represent the preprocessor expression which was parsed to obtain
+     * {@link getDirectFeatureMapping}. In either case, this label may be an arbitrary value,
+     * selected according to the needs of the user of this class.
+     */
+    public abstract List<String> getLabelLines();
+
+    /**
+     * Returns the start line number of the {@link getLabelLines label} of this node in the
+     * corresponding source code.
+     *
+     * @see setFromLine
+     */
+    public abstract int getFromLine();
+
+    /**
+     * Changes the line number returned by {@link getFromLine}.
+     *
+     * @see getFromLine
+     */
+    public abstract void setFromLine(int from);
+
+    /**
+     * Returns the end line number of the {@link getLabelLines label} of this node in the
+     * corresponding source code.
+     *
+     * <p>The line number is exclusive (i.e., it points 1 behind the last included line).
+     *
+     * @see setToLine
+     */
+    public abstract int getToLine();
```

Comment:
For consistency with `DiffLineNumber` and `Lines`, this should probably be called `getToLineExclusive`. It's a bit verbose but I hope that this helps to avoid some indexing bugs for users right away. Please rename this method and all the other methods acting on line numbers in this class where appropriate.

Reply:
See below. I used `Lines` to get rid of all these single line number functions.

Reply:
See this commit: eb2134152e184d7bd2c245a0f132141e2f2d5eec